src/twister2/builder/smart_builder_plugin.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+
+import json
+import logging
+import os
+import tempfile
+import time
+from pathlib import Path
+
+import pytest
+from filelock import FileLock
+
+from twister2.builder.builder_abstract import BuilderAbstract
+
+_TMP_DIR: str = tempfile.gettempdir()
+BUILD_STATUS_FILE_NAME: str = 'twister_builder.json'
+BUILD_LOCK_FILE_PATH: str = os.path.join(_TMP_DIR, 'smart_builder.lock')
+
+logger = logging.getLogger(__name__)
+
+
+class BuildStatus:
+    NOT_DONE = 0
+    IN_PROGRESS = 1
+    DONE = 2
+    FAILED = 3
+
+
+class SmartBuilderPlugin:
```

Comment:
I know, that probably this name was suggested by my own, but what do you think about change name from `SmartBuilder` into `BuildManager`? Maybe it better reflects what this class really do. Moreover, in `builder.py` fixtures it will be simpler to distinguish this plugin from builder like `WestBuilder`. Also, in this class in method `build` there is variable `builder` and for me it is a little bit tricky, that user has to provide `builder` into `SmartBuilder` class. But this is only my opinion - we can discuss about it.

Reply:
I made it as a plugin because it was the simplest way to make sure clear() method will be  called only once while we use xdist. Maybe there is better solution to solve that problem but I didn't find. You cannot use singleton pattern because each worker is run as separated process (with own instance of session). 
The name of this class doesn't matter, we can remove `plugin` from the the name. Also we can provide some pytest hook to allow users to change this builder manager to something different.